### ISSUE TITLE
php open/close tags ignored within C-style comments

### ISSUE THREAD

**Issue by Daniel-Peder** [php/doc-en#1355]
### Description

[The following code:

```php
<?php

echo (ob_start()?'':'') 
. 
/* 
?> 
hello world 
<?php 
*/ 
ob_get_clean()
,
PHP_EOL
;
```

Resulted in this output:
```

```

But I expected this output instead:
```
hello world
```
](https://3v4l.org/OkFOE)

The open/close tags <?php ... ?> should have precedence over the comments or at least, it should be better described in manual page.

According to the [php tags manual page](https://www.php.net/manual/en/language.basic-syntax.phptags.php), telling:
> When PHP parses a file, it looks for opening and closing tags, which are <?php and ?> which tell PHP to start and stop interpreting the code between them. 

PHP does not obey this rule, as it rather considers the text between tags as part of comment, which is, however, rather part of interpreted php code instead of part of not interpreted text.

### PHP Version

ALL

**Comment by KapitanOczywisty** [comment]
1. I've mentioned the exception when closing tag has precedence
2. It might be poorly documented, but it's consistent with how php works.
3. If that was the case it would be syntax error anyway, but let say it would work like brackets:

```php
<?php
if($a){
?>Hello world<?php
}
```
is becoming:
```php
<?php
if($a){
echo 'Hello world';
}
```
So taking the same principles:
```php
<?php
/*
?>Hello world<?php
*/
```
Would become
```php
<?php
/*
echo 'Hello world';
*/
```
What purpose such code would serve?

**Comment by Daniel-Peder** [comment]
> ... and that isn't causing any actual problems for anyone.

It is causing probles, at least to me.
The motivation is to be able to pass html code as argument to functions while being able still use the IDE code highlighting and formatting capabilites, like:

```php
function my_example_html_echo ( $html ) {
  echo $html
}

my_example_html_echo ( ob_start_returning_empty_string() . /*
?>
<div>
<p>my formatted html code</p>
</div>
<?php */ ob_get_clean()
);

```

**Comment by Daniel-Peder** [comment]
> What purpose such code would serve?

see my reply above

**Comment by damianwadley** [comment]
> The motivation is to be able to pass html code as argument to functions while being able still use the IDE code highlighting and formatting capabilites, like:

No offense, but that's a hack for an IDE that doesn't support highlighting inside PHP strings. Not that I know of any which can do that intelligently, but fighting with the PHP syntax to get something that works isn't the answer.

There is a way to do this with minimal additional syntax: use a function. https://3v4l.org/pYbjn

Honestly, I'd rather see a convention with heredoc/nowdoc where the identifier is a language name, like "HTML" or "JAVASCRIPT", that hints to editors how it should be highlighted.
```php
$html = <<<HTML
<p>This could also support $variables without having to <code>use</code> them in a closure.</p>
HTML;
```

**Comment by Daniel-Peder** [comment]
> 
> There is a way to do this with minimal additional syntax: use a function. https://3v4l.org/pYbjn

sure, I know the way to use closure, however, it is adding unnecessary complexity

Also no offence, the ability of IDE to highlight or not highlight html withing string is just flamewar starter, not argument. I just still insist on documented behaviour and my real needs in real IDE to use such documented behaviour. Not to be forced to change my working habits just because things are not working as documented.

**Comment by cmb69** [comment]
If things are not working as documented, that can be an implementation or a documentation bug. In this case it is clearly a documentation bug (although a minor one). If you like the behavior to change, consider to [pursue the RFC process](https://wiki.php.net/rfc/howto).

**Comment by Daniel-Peder** [comment]
> If things are not working as documented, that can be an implementation or a documentation bug. In this case it is clearly a documentation bug (although a minor one). If you like the behavior to change, consider to [pursue the RFC process](https://wiki.php.net/rfc/howto).

Thanks for advice. From my perspective this is a bug. The open/close tags should have absolute precedence over all other language constructs and should immediately start/abort php processing as declared:
> Parsing in this manner allows PHP to be embedded in all sorts of different documents, as everything outside of a pair of opening and closing tags is ignored by the PHP parser. 
https://www.php.net/manual/en/language.basic-syntax.phptags.php

**Comment by cmb69** [comment]
The manual also [states elsewhere](https://www.php.net/manual/en/language.basic-syntax.comments.php)

> 'C' style comments end at the first */ encountered.

And that is what is parsed here: open PHP tag, start of comment, end of comment.

**Comment by KapitanOczywisty** [comment]
> The motivation is to be able to pass html code as argument to functions while being able still use the IDE code highlighting and formatting capabilites, like:

As far I know, most IDE supports this kind of syntax:

```php
my_example_html_echo( do_some_operations_on_html(<<<HTML
<div>
<p>my formatted and possibly highligthed html code</p>
</div>
HTML ));
```

And if you want to deal with `ob_start` or IDE have troubles with heredocs, why not use output buffering without hacking?
```php
ob_start();
?>
<div>
<p>my formatted and possibly highligthed html code</p>
</div>
<?php
my_example_html_echo(ob_get_clean());
```

**Comment by Daniel-Peder** [comment]
> And if you want to deal with `ob_start` or IDE have troubles with heredocs, why not use output buffering without hacking?

Thanks. However, I dont need You to teach me php. 
The most close approach for my needs is using closures, while it is still too complex, so i just tried to point out to that obvious bug. 
```php
my_example_html_echo(
  (function(){
    ob_start(); ?>
      <div>
        <p>my formatted and possibly highligthed html code</p>
      </div>
    <?php
    return ob_get_clean();
  })()
);
```

**Comment by iluuu1994** [comment]
You're abusing comments in your original example just because it produces a syntax error without them. That doesn't mean the right solution is to make PHP execute parts of the comment.

Note that `?> ... <?php` gets converted to `echo ' ... ';` under the hood (https://3v4l.org/WTYSo/vld). As this is a statement it doesn't make sense to use it in an expression. You're also abusing string concatenation in your last example just to wrap three statements together (start, echo, stop). One solution is to move the code out of the function call and keep them as separate statements.

```php
my_ob_html_open();
?>
    <div>
        <p>my formatted and possibly highligthed html code</p>
    </div>
<?php 
$result = my_ob_html_close();
my_example_html_echo($result);
```

This is one approach but it's hardly elegant. The best approach was already described by @KapitanOczywisty, use a string, in whatever form you prefer. That also avoid unnecessary overhead of writing to the buffer and clearing it right after.

**Comment by Daniel-Peder** [comment]
> Note that `?> ... <?php` gets converted to `echo ' ... ';` under the hood (https://3v4l.org/WTYSo/vld). 

Please, You note, what is the doc saying:

> Parsing in this manner allows PHP to be embedded in all sorts of different documents, as everything outside of a pair of opening and closing tags is ignored by the PHP parser.
https://www.php.net/manual/en/language.basic-syntax.phptags.php

note the : *ignored by the PHP parser* - hwoever, if it worked generic way as stated, it would just flush out evertyhing that follows the closing `?>` tag. and it is not my business to think what do they do under the hood. 

From my perspective, it just does not do what is stated.

Also I agree, Your example is exactly as You say 'hardly elegant'. This is why I do not like it too. Then why should we talk about it again?

**Comment by iluuu1994** [comment]
> Please, You note, what is the doc saying:
>
>> Parsing in this manner allows PHP to be embedded in all sorts of different documents, as everything outside of a pair of opening and closing tags is ignored by the PHP parser.
https://www.php.net/manual/en/language.basic-syntax.phptags.php

Technically still correct, as `/* ?> */` results in a single token.

https://3v4l.org/qYF1I#v8.1.2

> And sorry, but it is not my business to think what do they do under the hood. From my perspective, it just does not do what is written in the doc.

That's why this issue wasn't just closed. Clarifying how it works in the docs makes sense, changing the behavior and breaking millions of code bases clearly does not.

> Also I agree, Your example is exactly as You say 'hardly elegant'. This is why I do not like it too. Then why should we talk about it again?

We offered alternatives that work. If you can come up with a better solution, take the time to go through the RFC process and provide an implementation, feel free to do so.

**Comment by Daniel-Peder** [comment]
> Technically still correct, as `/* ?> */` results in a single token.

Nope, technically not correct, as the parser should stop parsing after closing tag. Thus should ignore the closing comment until reaching the next opening tag `<?php`.

Which is what I wrote before, it does not what promissed. It did not stopped parsing. Also technically, the comment is still part of code. So it was parsed and it should not.

**Comment by Daniel-Peder** [comment]
> We offered alternatives that work. 
I was not asking for alternatives that work. I got my own alternatives, which I mentioned. The closure. Also is provided explanation, why it not good enough. Otherwise I can find my solutions how to do other php ways. Including these You provided. Just these are not good enough to me.

**Comment by emptyheap2019** [comment]
@Daniel-Peder I can agree with you that there is specifically one part of the documentation that could use improvement.

Specifically...

> Parsing in this manner allows PHP to be embedded in all sorts of different documents, as everything outside of a pair of opening and closing tags is ignored by the PHP parser.
> https://www.php.net/manual/en/language.basic-syntax.phptags.php

In that, the correct (well-defined) behaviour is more technically illustrated as the closing tag exits the parser and every proceeding token is defined as a string, up untill the point that an `<?php` opening tag is discovered and the parser will return from a _NOOP_. Which means that if the parser defines a comment as __not code__ [SIC] then there is no way to circumvent identifying the closing tag (i.e. code) inside of that comment. This is because comments are intended to act as a way of seperating code from data. Now the purpose of doing this is that, while written in code, the comment is actually only useful to the human that is reading it and not the parser. So right away I can't agree with your thesis statement that there could be better way to do what you want.

However what I can offer you is sprucing up the language in the documentation a bit to elaborate on how the openning/closing tags effect the parser (which is definitley __well__ defined behaviour). Every other part of the docs aforementioned, however, looks clear to me (comments included).

**Comment by Daniel-Peder** [comment]
> @Daniel-Peder I can agree with you that there is specifically one part of the documentation that could use improvement.
> 

@emptyheap2019 yep, thanks, I understand the mechanics, however, I just wanted to raise the case. 

There would be need to add some logic while the parser crawls trough the comment data looking not only fo the closing comment token but also for php closing tag.

```
/* 
?> 
any text (or html) data to feed into output buffer
<?php 
<=== at this point the parser should remember it left with comment opened and continue in comment "mode", not parsing the php code
*/
```

Yep. From previous discussion I understood, that there should be someone to write that code, who is surely not me and there could be a big impact to backward compatibility. Maybe some ini option could be considered in the future. 

Until then, I will stick with yet available coding ways. 

Thanks.

**Comment by Daniel-Peder** [comment]
... or to add two new tokens, `/* ?>` and `<?php */` with the extra functionality as discussed above, not enabling any other comments nor any additional logic ... yeah, just idea
